Cabal/Distribution/Simple/Glob.hs

```diff
@@ -86,35 +106,59 @@ explainGlobSyntaxError filepath VersionDoesNotSupportGlob =
 
 data IsRecursive = Recursive | NonRecursive
 
+data MultiDot = MultiDotDisabled | MultiDotEnabled
+
 data Glob
-  = GlobStem String Glob
+  = GlobStem FilePath Glob
     -- ^ A single subdirectory component + remainder.
   | GlobFinal GlobFinal
 
 data GlobFinal
-  = FinalMatch IsRecursive String
+  = FinalMatch IsRecursive MultiDot String
     -- ^ First argument: Is this a @**/*.ext@ pattern?
-    --   Second argument: the extensions to accept.
+    --   Second argument: should we match against the exact extensions, or accept a suffix?
+    --   Third argument: the extensions to accept.
   | FinalLit FilePath
     -- ^ Literal file name.
 
-fileGlobMatches :: Glob -> FilePath -> Bool
-fileGlobMatches pat = fileGlobMatchesSegments pat . splitDirectories
+-- | Returns 'Nothing' if the glob didn't match at all, or 'Just' the
+--   result if the glob matched (or would have matched with a higher
+--   cabal-version).
+fileGlobMatches :: Glob -> FilePath -> Maybe (GlobResult FilePath)
+fileGlobMatches pat candidate = do
+  match <- fileGlobMatchesSegments pat (splitDirectories candidate)
+  return (candidate <$ match)
 
-fileGlobMatchesSegments :: Glob -> [FilePath] -> Bool
-fileGlobMatchesSegments _ [] = False
+fileGlobMatchesSegments :: Glob -> [FilePath] -> Maybe (GlobResult ())
+fileGlobMatchesSegments _ [] = Nothing
 fileGlobMatchesSegments pat (seg : segs) = case pat of
-  GlobStem dir pat' ->
-    dir == seg && fileGlobMatchesSegments pat' segs
+  GlobStem dir pat' -> do
+    guard (dir == seg)
+    fileGlobMatchesSegments pat' segs
   GlobFinal final -> case final of
-    FinalMatch Recursive ext ->
+    FinalMatch Recursive multidot ext -> do
       let (candidateBase, candidateExts) = splitExtensions (last $ seg:segs)
-      in ext == candidateExts && not (null candidateBase)
-    FinalMatch NonRecursive ext ->
+      guard (not (null candidateBase))
```

Comment:
So `*.html` won't match `.foo.html`. Makes sense, since that's how it works in shell, but we should document it.